framework/decode/dx12_replay_consumer_base.h

```diff
@@ -163,15 +163,51 @@ class Dx12ReplayConsumerBase : public Dx12Consumer
     }
 
     template <typename T>
-    void AddObject(const format::HandleId* p_id, T** pp_object)
+    void SetObjectName(const format::HandleId* p_id, T** pp_object, format::ApiCallId call_id)
+    {
+        if ((p_id != nullptr) && (pp_object != nullptr) && (*pp_object != nullptr))
+        {
+            // Don't override object name if this AddObject() call is triggered by a QueryInterface() call
+            if (call_id != format::ApiCall_IUnknown_QueryInterface)
+            {
+                IUnknown* iunknown = reinterpret_cast<IUnknown*>(*pp_object);
+                ID3D12Object* object = nullptr;
+
+                // See if this is a D3D12Object
+                if (SUCCEEDED(iunknown->QueryInterface(IID_ID3D12Object, reinterpret_cast<void**>(&object))))
+                {
+                    const std::wstring constructed_name = ConstructObjectName(*p_id, call_id);
+
+                    HRESULT res = object->SetName(constructed_name.c_str());
+                    GFXRECON_ASSERT(res == S_OK);
+
+                    // QueryInterface() increases the refcount on success, so release here
+                    iunknown->Release();
```

Comment:
Should `Release()` be called on `object` instead of `iunknown`? Or maybe they're guaranteed to be the same object so it doesn't matter?

Alternatively a smart com ptr typedef `ID3D12ObjectComPtr` could be added to `dx12_util.h` and used to release automatically.

Reply:
Done